### PR TITLE
Bitmask Arrays

### PR DESCRIPTION
**Pull request by kbwestfall**
Prompted by @jhennawi's masking questions, I've implemented a class that hopefully makes the use of the bitmasks easier.  So far, I've only implemented a new class (`BitMaskArray`) and put in some docs and tests.  Before I propagate this through the code, please comment on the changes so far.  In particular, look at

- `doc/include/bitmaskarray_usage.rst`: This is documents how to use the new class
- `pypeit/tests/test_pypeitimage.py`: Has new tests that show the functionality
- `pypeit/images/bitmaskarray.py`: The implementation of the base class

### CODE REVIEW

**Review by jhennawi — APPROVED** [review]
Hi @kbwestfall thanks for this PR and for making this more clear. The only substantive comments I have are: 

1) My main concern was how to create an PypeItImage from an existing bad pixel mask (which does not have bits specified, i.e. a boolean). This is being done i.e. here: 

https://github.com/pypeit/PypeIt/blob/develop/pypeit/coadd2d.py (line 566) 

and here: 

https://github.com/pypeit/PypeIt-development-suite/blob/develop/dev_algorithms/jwst/jwst_diff_coadd2d.py (line 380)

Is what is being done there correct? 

2) Also, is it possible to make the PypeItImage API clearer  on how one can instantiate from an existing image. Right now  every argument there is optional and there are no examples on how one should deal with this usage case, and specifically on how to instanatiate from a mask.

**Comment by kbwestfall** [comment]
> Hi @kbwestfall thanks for this PR and for making this more clear. The only substantive comments I have are:
> 
>     1. My main concern was how to create an PypeItImage from an existing bad pixel mask (which does not have bits specified, i.e. a boolean). This is being done i.e. here:
> 
> 
> https://github.com/pypeit/PypeIt/blob/develop/pypeit/coadd2d.py (line 566)
> 
> and here:
> 
> https://github.com/pypeit/PypeIt-development-suite/blob/develop/dev_algorithms/jwst/jwst_diff_coadd2d.py (line 380)
> 
> Is what is being done there correct?
> 
>     2. Also, is it possible to make the PypeItImage API clearer  on how one can instantiate from an existing image. Right now  every argument there is optional and there are no examples on how one should deal with this usage case, and specifically on how to instanatiate from a mask.

Thanks for having a look, @jhennawi .  This introduction of the bitmask array is the first step in addressing both of these.  I'll keep going so you can see how this affects the PypeItImage datamodel and use.  Stay tuned...

**Review by profxj — APPROVED** [review]
Super helpful, thanks!

**Comment by kbwestfall** [comment]
I've now propagated the bitmaskarray use to PypeItImage and the spec2d object.  Unit tests pass, and I've started the dev-suite running.

**Comment by kbwestfall** [comment]
```
Test Summary
--------------------------------------------------------
--- PYTEST PYPEIT UNIT TESTS PASSED  204 passed, 707 warnings in 184.70s (0:03:04) ---
--- PYTEST UNIT TESTS PASSED  119 passed, 1544 warnings in 901.68s (0:15:01) ---
--- PYTEST VET TESTS PASSED  25 passed, 979 warnings in 767.36s (0:12:47) ---
--- PYPEIT DEVELOPMENT SUITE FAILED 1/154 TESTS  ---
Failed tests:
    gemini_gmos/GN_E2V_MULTI_R400_600 pypeit
Skipped tests:
Testing Started at 2022-11-19T03:42:18.547074
Testing Completed at 2022-11-19T11:30:43.064477
Total Time: 7:48:24.517403
```
I can't reproduce the test failure locally.  @badpandabear , @profxj , do you think this is the flexure issue you were seeing in other testing?

File: pypeit/alignframe.py [hunk 1]
```diff
@@ -157,7 +157,8 @@ def __init__(self, rawalignimg, slits, spectrograph, alignpar, det=1,
         self.slits = slits
 
         # Optional parameters
-        self.bpm = rawalignimg.mask if msbpm is None else msbpm
+#        self.bpm = rawalignimg.mask if msbpm is None else msbpm
```
Comment (profxj) [inline pypeit/alignframe.py]: remove?

File: pypeit/bitmask.py [hunk 2]
```diff
@@ -410,33 +410,67 @@ def unpack(self, value, flag=None):
...
-            if key == 'NULL':
-                continue
-            hdr['{0}{1}'.format(prefix, str(value).zfill(ndig))] = (key, self.descr[value])
+        bits = self.to_dict(prefix=prefix)
+        for hdr_key, bit_key in bits.items():
+            i = int(hdr_key.replace(prefix, ''))
+            hdr[hdr_key] = (bit_key, self.descr[i])
+#        if prefix is None:
+#            prefix = self.prefix
+#        maxbit = max(list(self.bits.values()))
```
Comment (profxj) [inline pypeit/bitmask.py]: remove?

File: pypeit/extraction.py [hunk 3]
```diff
@@ -806,8 +802,7 @@ def local_skysub_extract(self, global_sky, sobjs, spat_pix=None, model_noise=Tru
         # Set the bit for pixels which were masked by the extraction.
         # For extractmask, True = Good, False = Bad
         iextract = base_gpm & np.logical_not(self.extractmask)
-        # TODO: Change this to use the update_mask method?
-        self.outmask[iextract] = self.sciImg.bitmask.turn_on(self.outmask[iextract], 'EXTRACT')
+        self.outmask.turn_on('EXTRACT', select=iextract)
```
Comment (profxj) [inline pypeit/extraction.py]: Clean!

File: pypeit/images/bitmaskarray.py [hunk 4]
```diff
@@ -0,0 +1,293 @@
+"""
```
Comment (profxj) [inline pypeit/images/bitmaskarray.py:1]: Nice!

File: pypeit/images/pypeitimage.py [hunk 5]
```diff
@@ -176,23 +164,60 @@ def from_pypeitimage(cls, pypeitImage):
...
         args, _, _, values = inspect.getargvalues(inspect.currentframe())
-        super(PypeItImage, self).__init__(d={k: values[k] for k in args[1:]})
+        # Loop through them all and assign them directly to attributes of the
+        # class if they're part of the datamodel and *not* None.
+        for key in args[1:]:
+            if key in self.datamodel and values[key] is not None:
+                setattr(self, key, values[key])
+
+        # Make sure all the relevant arrays have the same shape
+        for attr in ['ivar', 'nimg', 'amp_img', 'det_img', 'rn2img', 'base_var', 'img_scale',
```
Comment (profxj) [inline pypeit/images/pypeitimage.py:187]: Is there a particular reason to move here (and out of `validate()`) ?

File: pypeit/images/pypeitimage.py [hunk 6]
```diff
@@ -658,74 +682,184 @@ def select_flag(self, flag=None, invert=False):
...
+            a boolean bad-pixel mask (or a good-pixel mask when ``invert`` is
+            True).  If ``flag`` is not provided, pixels flagged for any reason
+            are returned as True.
+        """
+        return self.fullmask.flagged(flag=flag, invert=invert)
+
+    def __sub__(self, other):
+        return self.sub(other)
+
+    def sub(self, other):
```
Comment (profxj) [inline pypeit/images/pypeitimage.py:695]: Thanks for cleaning this up!

File: pypeit/images/pypeitimage.py [hunk 7]
```diff
@@ -658,74 +682,184 @@ def select_flag(self, flag=None, invert=False):
...
+        if self.base_var is not None or other.base_var is not None:
+            new_base = np.zeros(newimg.shape) if self.base_var is None else self.base_var
+            if other.base_var is not None:
+                new_base += other.base_var
+        else:
+            new_base = None
+
+        # Image scaling
+        # TODO: This is bogus.  Maybe we should just set this to 1?  Either way,
+        # trying to model the noise using our current approach won't be
```
Comment (profxj) [inline pypeit/images/pypeitimage.py:764]: Good point..

File: pypeit/images/pypeitimage.py [hunk 8]
```diff
@@ -658,74 +682,184 @@ def select_flag(self, flag=None, invert=False):
...
-        #new_sciImg.crmask = crmask_diff | np.logical_not(outmask_comb)
-        # Note that the following uses the saturation and mincounts held in self.detector
-        new_sciImg.build_mask()
-
-        return new_sciImg
+            new_pypeitImage.files = self.files + other.files
+
+        return new_pypeitImage
+
+#    def sub(self, other, par):
```
Comment (profxj) [inline pypeit/images/pypeitimage.py]: Remove

File: pypeit/tests/test_pypeitimage.py [hunk 9]
```diff
@@ -29,8 +30,44 @@ def test_full():
...
     # Test
-    assert isinstance(_pypeitImage.image, np.ndarray)
-    assert _pypeitImage.ivar is None
+    assert np.array_equal(_pypeitImage.image, np.ones(shape)), 'image array changed'
+    assert np.array_equal(_pypeitImage.ivar, np.ones(shape)), 'ivar array changed'
+    assert isinstance(_pypeitImage.fullmask, imagebitmask.ImageBitMaskArray), 'mask type changed'
+    assert np.array_equal(_pypeitImage.fullmask.mask, pypeitImage.fullmask.mask), \
+                'mask array changed'
+
+def test_sub():
```
Comment (profxj) [inline pypeit/tests/test_pypeitimage.py:39]: Thx for the new unit tests!

**Review by profxj — APPROVED** [review]
Whoa!

Thanks for the deep dive.   Nothing too
substinative from me, I think.  And I think
I finally get the `bitmask`'ing :)

**Comment by kbwestfall** [comment]
Tests pass:
```
Test Summary
--------------------------------------------------------
--- PYTEST PYPEIT UNIT TESTS PASSED  204 passed, 707 warnings in 188.02s (0:03:08) ---
--- PYTEST UNIT TESTS PASSED  119 passed, 1544 warnings in 918.23s (0:15:18) ---
--- PYTEST VET TESTS PASSED  25 passed, 978 warnings in 787.72s (0:13:07) ---
--- PYPEIT DEVELOPMENT SUITE PASSED 154/154 TESTS  ---
Testing Started at 2022-11-30T19:06:57.980069
Testing Completed at 2022-12-01T03:05:30.701370
Total Time: 7:58:32.721301
```

**Pull request merged**